bin/server/cmd/root.go

```diff
@@ -232,162 +213,34 @@ func start() {
-}
-
-func initSetting(ctx context.Context, store *store.Store) (*config, error) {
-	result := &config{}
-	{
-		configCreate := &api.SettingCreate{
-			CreatorID:   api.SystemBotID,
-			Name:        api.SettingAuthSecret,
-			Value:       common.RandomString(secretLength),
-			Description: "Random string used to sign the JWT auth token.",
-		}
-		config, err := store.CreateSettingIfNotExist(ctx, configCreate)
-		if err != nil {
-			return nil, err
+		fmt.Printf("cannot new server, error: %v\n", err)
+		if s != nil {
+			_ = s.Shutdown()
 		}
-		result.secret = config.Value
-	}
-
-	return result, nil
-}
-
-func initBranding(ctx context.Context, store *store.Store) error {
-	configCreate := &api.SettingCreate{
-		CreatorID:   api.SystemBotID,
-		Name:        api.SettingBrandingLogo,
-		Value:       "",
-		Description: "The branding logo image in base64 string format.",
-	}
-	_, err := store.CreateSettingIfNotExist(ctx, configCreate)
-	if err != nil {
-		return err
-	}
-
-	return nil
-}
-
-// Run will run the main server.
-func (m *Main) Run(ctx context.Context) error {
-	ctx, cancel := context.WithCancel(ctx)
-	m.serverCancel = cancel
-
-	db, err := m.metadataDB.connect()
-	if err != nil {
-		return fmt.Errorf("cannot new db: %w", err)
-	}
-	if err := db.Open(ctx); err != nil {
-		return fmt.Errorf("cannot open db: %w", err)
-	}
-	m.db = db
-
-	cacheService := server.NewCacheService(m.l)
-	storeInstance := store.New(m.l, db, cacheService)
-
-	config, err := initSetting(ctx, storeInstance)
-	if err != nil {
-		return fmt.Errorf("failed to init config: %w", err)
-	}
-
-	err = initBranding(ctx, storeInstance)
-	if err != nil {
-		return fmt.Errorf("failed to init branding: %w", err)
-	}
-
-	s := server.NewServer(m.l, storeInstance, m.lvl, version, host, m.profile.Port, frontendHost, frontendPort, m.profile.DatastorePort, m.profile.Mode, m.profile.DataDir, m.profile.BackupRunnerInterval, config.secret, readonly, demo, debug)
-
-	s.ActivityManager = server.NewActivityManager(s, storeInstance)
-
-	licenseService, err := enterprise.NewLicenseService(m.l, m.profile.DataDir, m.profile.Mode)
-	if err != nil {
-		return err
+		return
 	}
-	s.LicenseService = licenseService
-	s.InitSubscription()
-
-	m.server = s
-
-	fmt.Printf(greetingBanner, fmt.Sprintf("Version %s has started at %s:%d", version, host, m.profile.Port))
 
+	// Execute program.
 	if err := s.Run(ctx); err != nil {
-		return err
-	}
-
-	return nil
-}
-
-// Close gracefully stops the program.
-func (m *Main) Close() error {
-	m.l.Info("Trying to stop Bytebase...")
-	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
-	defer cancel()
-
-	if m.server != nil {
-		m.l.Info("Trying to gracefully shutdown server...")
-		m.serverCancel()
-		m.server.Shutdown(ctx)
-	}
-
-	if m.db != nil {
-		m.l.Info("Trying to close database connections...")
-		if err := m.db.Close(); err != nil {
-			return err
+		if err != http.ErrServerClosed {
+			logger.Error(err.Error())
+			if s != nil {
```

Comment:
We have lots of places doing s.Shutdown(). We can consider doing in a deferred function. However, we need to think of a way to coordinate with the go routine in line218.

Reply:
Yes, the core problem is that how we shutdown gracefully if receive SIGKILL/SIGTERM in NewServer(). We need close driver, connection and Postgres server(if embed). I had tried defer, but still not ideal.